File: Flix.Backend/Flix.Services/Database/Activity.cs
using Flix.Model.Enums;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Flix.Services.Database
{
    public class Activity
    {
        [Key]
        public int Id { get; set; }
        public int UserId { get; set; }
        [ForeignKey(nameof(UserId))]
        public User User { get; set; } = null!;

        public ActivityType Type { get; set; }

        public int? MovieId { get; set; }
        [ForeignKey(nameof(MovieId))]
        public Movie? Movie { get; set; }

        public int? ReviewId { get; set; }
        [ForeignKey(nameof(ReviewId))]
        public Review? Review { get; set; }

        public int? ClashId { get; set; }
        [ForeignKey(nameof(ClashId))]
        public Clash? Clash { get; set; }

        public int? MovieListId { get; set; }
        [ForeignKey(nameof(MovieListId))]
        public MovieList? MovieList { get; set; }

        public int? TargetUserId { get; set; }
        [ForeignKey(nameof(TargetUserId))]
        public User? TargetUser { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }
}

File: Flix.UI/flix_desktop/lib/models/activity.dart
import 'package:flix_desktop/enums/activity_type.dart';
import 'package:flix_desktop/models/clash.dart';
import 'package:flix_desktop/models/movie.dart';
import 'package:flix_desktop/models/movie_list.dart';
import 'package:flix_desktop/models/review.dart';
import 'package:flix_desktop/models/user.dart';
import 'package:json_annotation/json_annotation.dart';

part 'activity.g.dart';

@JsonSerializable()
class Activity {
  Activity(
      this.id,
      this.user,
      this.type,
      this.movie,
      this.review,
      this.clash,
      this.movieList,
      this.targetUser,
      this.createdAt
      );

  final int? id;

  final User? user;

  final ActivityType? type;

  final Movie? movie;
  final Review? review;
  final Clash? clash;
  final MovieList? movieList;

  final User? targetUser;

  final DateTime? createdAt;

  Movie? get relatedMovie => movie ?? review?.movie;

  factory Activity.fromJson(Map<String, dynamic> json) =>
      _$ActivityFromJson(json);

  Map<String, dynamic> toJson() => _$ActivityToJson(this);
}
